### PR TITLE
Add CSV upload component for Step 2 of the Plan Wizard

### PR DESCRIPTION
**Pull request by mturley**

### CODE REVIEW

**Comment by mturley** [comment]
I'm integrating this with a custom redux-form `Field` component to add validation... will need some review tomorrow. As soon as I finish that and get any feedback on it, though, this one will be done. Next up is #13, putting this data together into API requests.

**Comment by mturley** [comment]
This could use some tests and such, and also I need a second opinion about the confirm/warning dialog behavior, (a) because I am using the native browser global `confirm()` as a placeholder (rather than go down a rabbit hole of patternfly-modal-inside-patternfly-modal), and (b) because there might a better design anyway. See my comment in #34.

Otherwise, though, this is ready to merge I think.. @AparnaKarve, even though I thought I hooked up the redux-form validator properly, I want it to disable the Next button until you import a file, and the button is enabled no matter what I do. So I need help figuring that part out.

**Comment by AparnaKarve** [comment]
>  I want it to disable the Next button until you import a file,

@mturley The below patch should keep Next button disabled until you import a file with at least 1 CSV row

```diff
diff --git a/app/javascript/react/screens/App/Overview/screens/PlanWizard/PlanWizardActions.js b/app/javascript/react/screens/App/Overview/screens/PlanWizard/PlanWizardActions.js
index c6859f6..5cf04b4 100644
--- a/app/javascript/react/screens/App/Overview/screens/PlanWizard/PlanWizardActions.js
+++ b/app/javascript/react/screens/App/Overview/screens/PlanWizard/PlanWizardActions.js
@@ -13,4 +13,5 @@ export const planWizardExitedAction = () => dispatch => {
   });
   // Dispatch reset for all the wizard step forms here
   dispatch(reset('planWizardGeneralStep'));
+  dispatch(reset('planWizardCSVStep'));
 };
```

```diff
diff --git a/app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/CSVDropzoneField.js b/app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/CSVDropzoneField.js
index 3a81a91..b98034e 100644
--- a/app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/CSVDropzoneField.js
+++ b/app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/CSVDropzoneField.js
@@ -53,6 +53,7 @@ class CSVDropzoneField extends React.Component {
 
   render() {
     const { input: { value } } = this.props;
+    const { meta: { pristine, error } } = this.props;
     const mainContents =
       !value || value.length === 0 ? (
         <EmptyState>
@@ -104,7 +105,8 @@ class CSVDropzoneField extends React.Component {
           </Toolbar.RightContent>
           <div className="form-group">{mainContents}</div>
           <Toolbar.Results>
-            {(value && value.length) || 0} {__('Items')}
+            {!error && value && sprintf(__('%s Items'), value.length)}
+            {pristine && error}
           </Toolbar.Results>
         </Toolbar>
       </Dropzone>
@@ -116,7 +118,8 @@ CSVDropzoneField.propTypes = {
   input: PropTypes.shape({
     value: PropTypes.arrayOf(PropTypes.array),
     onChange: PropTypes.func
-  })
+  }),
+  meta: PropTypes.object
 };
```

```diff
diff --git a/app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/PlanWizardCSVStep.js b/app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/PlanWizardCSVStep.js
index f07061c..9ef78af 100644
--- a/app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/PlanWizardCSVStep.js
+++ b/app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/PlanWizardCSVStep.js
@@ -1,6 +1,6 @@
 import React from 'react';
 import { Field, reduxForm } from 'redux-form';
-import { required } from 'redux-form-validators';
+import { length } from 'redux-form-validators';
 import CSVDropzoneField from './CSVDropzoneField';
 
 const PlanWizardCSVStep = () => (
@@ -9,8 +9,9 @@ const PlanWizardCSVStep = () => (
     <Field
       name="csvRows"
       component={CSVDropzoneField}
-      required
-      validate={[required({ msg: __('Required') })]}
+      validate={[
+        length({ min: 1, msg: __('At least one VM needs to be selected') })
+      ]}
     />
   </div>
 );

```

File: app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/CSVDropzoneField.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      >
+        <Toolbar>
+          <Toolbar.RightContent>
+            <Button
+              bsStyle="primary"
+              onClick={() => {
+                dropzoneRef.open();
+              }}
+            >
+              <Icon type="fa" name="upload" /> {__('Import')}
```
Comment (serenamarie125) [inline app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/CSVDropzoneField.js]: @mturley can you remove the icon from the button, and change the button label to Browse?

Comment (mturley) [reply app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardCSVStep/CSVDropzoneField.js]: Sure. Was originally just copying the mock, sorry.

File: app/javascript/react/screens/App/Overview/components/ModalWizard/ModalWizard.js [hunk 2]
```diff
@@ -3,14 +3,18 @@ import PropTypes from 'prop-types';
...
 import { connect } from 'react-redux';
 
+// TODO we should lift this application-specific stuff out of this generalized component file.
 const reduxFormMap = {
   [__('Infrastructure Mapping Wizard')]: [
     'mappingWizardGeneralStep',
     'mappingWizardClustersStep'
   ],
-  [__('Migration Plan Wizard')]: []
+  [__('Migration Plan Wizard')]: ['planWizardGeneralStep']
```
Comment (AparnaKarve) [inline app/javascript/react/screens/App/Overview/components/ModalWizard/ModalWizard.js]: @mturley Add `planWizardCSVStep` here in `reduxFormMap` for Next to work.
Related to https://github.com/priley86/miq_v2v_ui_plugin/pull/64

**Comment by mturley** [comment]
@priley86, @AparnaKarve, this one should be all ready to go.

**Comment by priley86** [comment]
I will take a look first thing tomorrow!

**Comment by AparnaKarve** [comment]
@mturley Can you attach a sample VM CSV file in the PR that we could use for testing?
Thanks.

**Comment by mturley** [comment]
@AparnaKarve actually, any CSV will work fine... I didn't make any assumptions yet about how we use the CSV data, all rows and columns are persisted to redux. I've been testing with this one. 
[Untitled spreadsheet - Sheet1.csv.zip](https://github.com/priley86/miq_v2v_ui_plugin/files/1760437/Untitled.spreadsheet.-.Sheet1.csv.zip)

**Comment by AparnaKarve** [comment]
> actually, any CSV will work fine

True -- just wanted to know if there needs to be any other column other than VM names and id's

**Comment by AparnaKarve** [comment]
@mturley Per this mockup VM screen below, we would need the CSV to have VM names.
But for the API POST required in Step3, we would also need VM id's to be included in the CSV.

screen shot 2018-02-26 at 2 33 15 pm

The id's however, should not be displayed in the UI if possible.

**Comment by priley86** [comment]
per @mturley's request... going to go ahead and merge this PR and rebase it. @mturley - please just follow up on the issues and note anything remaining.

**Comment by mturley** [comment]
@AparnaKarve, ok thanks for the info. I figured we should merge this now to unblock some refactoring Patrick and I will be doing tonight, and then yeah as we get the POST data together we can nail down how the CSV data is handled.

**Comment by AparnaKarve** [comment]
@mturley I'm thinking, the CSV needs to look something like this -
```csv
10000000000001, bostonVM1
10000000000002, raleighVM2
10000000000003, westfordVM3
```

These id's would then be used for POSTing.

Once imported, the UI should *only* display the VM names (not the id's)
Currently, it is (partially) showing the id's -

screen shot 2018-02-26 at 9 25 26 pm

**Comment by mturley** [comment]
Sure, okay. I can have it exclude the first column. Maybe we should tell the user that the first and second columns should be id and name, respectively? I also wonder if we need to handle the case of having a header row at the top of the CSV, where the first row is not a VM but a title of each column.

**Comment by mturley** [comment]
@AparnaKarve do you think it's sufficient for us to specify/assume the structure of the CSV? After all, we give them no instructions beyond "a CSV of VMs". I wonder if we need to have some kind of selection where you tell it which column is the id and which is the name? We could just place an edit icon that reveals a button on each column header for that, if we want to. I am also curious if this will usually be a user generated CSV or something exported from some other tool. @serenamarie125, can you give any more context on the usability of this CSV step?

**Pull request merged**